File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []

        pairs = {
            ')' : '(',
            '}' : '{',
            ']' : '['
        }

        for ch in s:
            # opeming bracket
            if ch in "({[":
                stack.append(ch)

            else:
                if not stack:
                    return False

                if stack[-1] != pairs[ch]:
                    return False
                stack.pop()

        return len(stack) == 0

        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
    stack<char>st;
    for(int i = 0; i < s.size(); i++) {
        if(s[i] == '(' || s[i] == '{' || s[i] == '[') {
            st.push(s[i]);
        }

        if(s[i] == ')') {
            if(!st.empty() && st.top() == '(') {
                st.pop();
            } else {
                return false;
            }
        }

        if(s[i] == '}') {
            if(!st.empty() && st.top() == '{') {
                st.pop();
            } else {
                return false;
            }
        }

        if(s[i] == ']') {
            if(!st.empty() && st.top() == '[') {
                st.pop();
            } else {
                return false;
            }
        }
    } 

    if(st.empty()) {
        return true;
    } else {
        return false;
    }
    }
};